### ISSUE TITLE
New Middleware Proposal: httpContentEncoding (deflate, gzip, br, zstd)

### ISSUE THREAD

**Issue by willfarrell** [middyjs/middy#186]
A middleware to compress responses.

## options
- encoding (gzip, br, zstd) - what content encoding to use
- mime (\*/\*) - which mine types to apply encoding too
- minimumCompressionSize (0) - at what size to start compressing body
- level - at what level to compress at
- bodyType (string,stream) - use stream mode, ignores minimumCompressionSize

## after
- [x] Convert body to binary
- [x] Attach header `Content-Encoding`: gzip

Ideal approach being a port of [`compression`](https://github.com/expressjs/compression)

Serverless approach: https://www.npmjs.com/package/serverless-content-encoding

```js
const { Readable, Transform, PassThrough } = require('stream')
//import {pipeline} from 'stream/promises'  // available in node >=15
const {promisify} = require('util')
const {pipeline: pipelineCallback} = require('stream')
const pipeline = promisify(pipelineCallback)
const { createBrotliCompress, createGzip, createDeflate, constants } = require('zlib')

const transformStreams = {
  // zstd: (opt = {}) => new PassThrough(),
  br: (opts = {
    params: {
      [constants.BROTLI_PARAM_MODE]: constants.BROTLI_MODE_TEXT,
      [constants.BROTLI_PARAM_QUALITY]: constants.BROTLI_MAX_QUALITY,
      // [zlib.constants.BROTLI_PARAM_SIZE_HINT]: 0
    }
  }) => createBrotliCompress(opts),
  gzip: (opts = {}) => createGzip(opts),
  deflate: (opts = {}) => createDeflate(opts),
}

const defaults = {
  //zstdOptions: undefined,
  brOptions: undefined,
  gzipOptions: undefined,
  deflateOptions: undefined
}

const httpContentEncodingMiddleware = (opts) => {
  const options = { ...defaults, ...opts }

  const httpContentEncodingMiddlewareAfter = async (request) => {
    const { event, response } = request
    if (!event.preferredEncoding || typeof response.body !== 'string') { return }

    const readStream = Readable.from(response.body)
    
    let contentEncodingStream, contentEncoding
    for(const encoding of ['br','gzip','deflate']) {
      if (!event.preferredEncodings.includes(encoding)) continue
      contentEncodingStream = transformStreams[encoding](options[`${encoding}Options`])
      contentEncoding = encoding
      break;
    }
    if (!contentEncoding) return

    const chunks = []
    const writeStream = new Transform({
      transform(chunk, encoding, callback) {
        chunks.push(chunk)
        callback(null, null)
      }
    })

    await pipeline([
      readStream,
      contentEncodingStream,
      writeStream
    ])

    const compressed = Buffer.concat(chunks).toString('base64')
    // Only apply encoding if it's smaller
    if (0 < compressed.length && compressed.length < response.body.length) {
      response.headers['Content-Encoding'] = contentEncoding
      response.body = compressed
      response.isBase64Encoded = true
    }
  }

  return {
    after: httpContentEncodingMiddlewareAfter
  }
}

module.exports = httpContentEncodingMiddleware
```

## Refs
- https://forum.serverless.com/t/gzip-response-encoding/1271/7
- https://techblog.commercetools.com/gzip-on-aws-lambda-and-api-gateway-5170bb02b543

Please 👍🏻 this issue if it would have value to you

**Comment by vladgolubev** [comment]
That's cool!

By the way, would be nice to have Brotli as well. I compressed [LibreOffice with Brotli](https://github.com/vladgolubev/aws-lambda-libreoffice) recently and it helped to save 34 MB. But I understand it may be more involved - binary for Lambda, or slower JS implementation.

**Comment by vladgolubev** [comment]
Also, doesn't API GW already provide GZIP out of the box? 

https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-gzip-compression-decompression.html

**Comment by lmammino** [comment]
I'd love this :)
If I am getting your idea correctly, you want to use it only to provide a compressed response. Would it make sense to consider a middleware for compressed requests too (not sure if ApiGateway will handle that transparently)?

Also, regarding supporting other compression algorithms, we can maybe have different middlewares and make them work only based on the requested `Accept-Encoding` (which we would already have from the content negotiation middleware).

It might not be super easy to come up with a flexible solution, but I'd strive for composability over bulky middlewares with a lot of dependencies where possible.

**Comment by willfarrell** [comment]
<img width="855" alt="screen shot 2018-07-27 at 19 14 29" src="https://user-images.githubusercontent.com/1097020/43351572-62cd5474-91d1-11e8-8f35-b2a305367ead.png">

**Comment by willfarrell** [comment]
Closing because there is a solution from AWS. It would be nice if they supported `br` as well, but I expect in time they will. Please post if you feel there is value in creating a compression middleware. This could be a nice addition to v2 if there is a use case.

**Comment by nponeccop** [comment]
Note that brotli is supported almost universally: https://www.caniuse.com/brotli and it's almost always better than gzip, because it's much faster.

Also, `zstd` is already standardized at https://tools.ietf.org/html/rfc8878. It lacks browser support. but the nature of lambda is that it might work in internal glue and other scenarios where clients are as compliant as you make them. And zstd is a revolution, it's _both_ faster and compresses better than `gzip`, and its faster compression modes are competitive with `brotli`

So my point is to support all the 3: gzip, br and zstd. With 3 different middlewares or with a pluggable compression method, I guess.

**Comment by willfarrell** [comment]
I've put together a PoC for a `httpContentEncoding` middleware at the top. Would love some feedback.

**Comment by willfarrell** [comment]
PR created! https://github.com/middyjs/middy/pull/702
